From 257e8fc7133a4c074c5b2c93fa34246ad21eadb5 Mon Sep 17 00:00:00 2001
From: Michael Mann <mmann78@netscape.net>
Date: Tue, 14 Jul 2015 21:34:29 -0400
Subject: [PATCH 11/12] Openflow - prevent infinite loop

Change-Id: I93ce7151467c890c12f7d612b5a7eecf5f91c189
Ping-Bug: 11358
Reviewed-on: https://code.wireshark.org/review/9640
Reviewed-by: Alexis La Goutte <alexis.lagoutte@gmail.com>
Petri-Dish: Alexis La Goutte <alexis.lagoutte@gmail.com>
Tested-by: Petri Dish Buildbot <buildbot-no-reply@wireshark.org>
Reviewed-by: Michael Mann <mmann78@netscape.net>
(cherry picked from commit 828358d22c6bcf0a1ade5b3ffaa8018a385bfc6c)
Reviewed-on: https://code.wireshark.org/review/9643
(cherry picked from commit 6cb9d38f200e2ef9109479135a94fba11b5a1a77)
Reviewed-on: https://code.wireshark.org/review/10587
Reviewed-by: Balint Reczey <balint@balintreczey.hu>
---
 epan/dissectors/packet-openflow_v5.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/epan/dissectors/packet-openflow_v5.c b/epan/dissectors/packet-openflow_v5.c
index 1a8c660..9932e27 100644
--- a/epan/dissectors/packet-openflow_v5.c
+++ b/epan/dissectors/packet-openflow_v5.c
@@ -3443,6 +3443,7 @@ dissect_openflow_tablemod_v5(tvbuff_t *tvb, packet_info *pinfo _U_, proto_tree *
 {
     proto_item *ti;
     proto_tree *conf_tree;
+    int save_offset;
 
     /* uint8_t table_id; */
     if (tvb_get_guint8(tvb, offset) <= OFPTT_MAX) {
@@ -3466,7 +3467,12 @@ dissect_openflow_tablemod_v5(tvbuff_t *tvb, packet_info *pinfo _U_, proto_tree *
 
     /* struct ofp_table_mod_prop_header properties[0]; */
     while (offset < length) {
+        save_offset = offset;
         offset = dissect_openflow_tablemod_prop_v5(tvb, pinfo, tree, offset, length);
+        if (offset <= save_offset) {
+            /* We don't need to go backwards or introduce an infinite loop */
+            break;
+        }
     }
 }
 
-- 
2.1.4

